clang/test/SemaSYCL/deferred-diagnostics-emit.cpp

```diff
@@ -0,0 +1,104 @@
+// RUN: %clang_cc1  -fsycl -triple spir64 -fsycl-is-device -verify -fsyntax-only  %s
+//
+// Ensure that the SYCL diagnostics that are typically deferred, correctly emitted.
+
+// testing that the deferred diagnostics work in conjunction with the SYCL namespaces.
+inline namespace cl {
+namespace sycl {
+class queue {
+public:
+  template <typename T>
+  void submit(T CGF) {}
+};
+
+template <int I>
+class id {};
+
+template <int I>
+class range {};
```

Comment:
I do not see the purpose of defining these classes, it is unnecessary and makes test fat. Can we just use one function with `sycl_kernel` attribute? Only this is needed to show the compiler where the device code starts from.

Reply:
@Fznamznon 
We cannot just throw the sycl_kernel attribute on a function like we do in some of the other tests - what we are looking for, specifically, is that the deferred diagnostics are working with namespaces. 
Remember, there is that overhaul of the deferred diagnostic system which was merged and then backed out - that overhaul passed all existing tests but yet caused all of our SYCL deferred diagnostics to suddenly not work.  The purpose of this deferred-diagnostics-emit.cpp is to provide defensive testing that will fail if that overhaul makes the same mistake and is later re-introduced.   This code you are identifying came from the case where we isolated the problem when used against that overhaul. It has been confirmed against that. Before I can change it, I'll have to reconfirm. 

This PR has been open for over a week now, and is really quite small. This particular test file is defensive. I would prefer that we merge it because we all have other things demanding our attention. I can look into trimming this side test file, if you insist, but my preference is to merge this PR rather than having it languish another week. Let me know.

Reply:
I understand that you want to capture some problems and we won't forget about them, I appreciate it. But all that I'm asking is reduce the test case to have minimal reproducer for our problems. You want capture that we have problems with namespaces - use namespaces, instead of adding not necessary stuff. All these classes will appear in AST dumps for module and will make debugging harder.
BTW. We have tons of ugly/bad/unnecessary changes in SYCL just because someone was running out of time. So the fact that it's "long conversation"/"we don't have time", "someone very needed to merge it yesterday" etc, won't force me to give approve for the patch if this patch doesn't fix crash which happens in the compiler on every second code sample.

Reply:
done.